src/Compilers/CSharp/Portable/Symbols/SymbolWithAnnotations.cs

```diff
@@ -631,189 +610,83 @@ public TypeSymbolWithAnnotations SetUnknownNullabilityForReferenceTypes()
-            }
-
-            return new ReferenceTypeUnknownNullabilityWithCustomModifiers(typeSymbol, customModifiers);
+            return new NonLazyType(typeSymbol, isNullable: null, customModifiers);
         }
 
-        private class WithoutCustomModifiers : TypeSymbolWithAnnotations
+        private sealed class NonLazyType : TypeSymbolWithAnnotations
         {
-            protected readonly TypeSymbol _typeSymbol;
+            private readonly TypeSymbol _typeSymbol;
+            private readonly bool? _isNullable;
+            private readonly ImmutableArray<CustomModifier> _customModifiers;
 
-            public WithoutCustomModifiers(TypeSymbol typeSymbol)
+            public NonLazyType(TypeSymbol typeSymbol, bool? isNullable, ImmutableArray<CustomModifier> customModifiers)
             {
                 Debug.Assert((object)typeSymbol != null);
+                Debug.Assert(!customModifiers.IsDefault);
                 _typeSymbol = typeSymbol;
+                _isNullable = isNullable;
+                _customModifiers = customModifiers;
             }
 
             public sealed override TypeSymbol TypeSymbol => _typeSymbol;
-            public sealed override bool? IsNullable => _typeSymbol.IsNullableType();
-            public override ImmutableArray<CustomModifier> CustomModifiers => ImmutableArray<CustomModifier>.Empty;
+            public sealed override bool? IsNullable => _isNullable;
+            public override ImmutableArray<CustomModifier> CustomModifiers => _customModifiers;
 
             public override TypeSymbolWithAnnotations WithModifiers(ImmutableArray<CustomModifier> customModifiers)
             {
-                if (customModifiers.IsDefaultOrEmpty)
-                {
-                    return this;
-                }
-
-                return new WithCustomModifiers(_typeSymbol, customModifiers);
+                return new NonLazyType(_typeSymbol, _isNullable, customModifiers);
             }
 
             public override TypeSymbol AsTypeSymbolOnly() => _typeSymbol;
 
             // PROTOTYPE(NullableReferenceTypes): Use WithCustomModifiers.Is() => false
             // and set IsNullable=null always for GetTypeParametersAsTypeArguments.
-            public override bool Is(TypeSymbol other) => _typeSymbol.Equals(other, TypeCompareKind.CompareNullableModifiersForReferenceTypes);
+            public override bool Is(TypeSymbol other) => _typeSymbol.Equals(other, TypeCompareKind.CompareNullableModifiersForReferenceTypes) && _customModifiers.IsEmpty;
 
             protected sealed override TypeSymbolWithAnnotations DoUpdate(TypeSymbol typeSymbol, ImmutableArray<CustomModifier> customModifiers)
             {
-                return TypeSymbolWithAnnotations.Create(typeSymbol, customModifiers);
+                return new NonLazyType(typeSymbol, _isNullable, customModifiers);
             }
 
             public override TypeSymbolWithAnnotations AsNullableReferenceType()
```

Comment:
>AsNullableReferenceType [](start = 54, length = 23)

I wonder if this can be simplified. I think `AsNullableReferenceType` can be pulled into the base. Instead of checking for `_isNullable`, check for `IsNullable` (always true for the lazy nullable type).